kadena/solidity/contracts/src/Wrapper.sol

```diff
@@ -0,0 +1,164 @@
+pragma solidity ^0.8.25;
+
+import {console} from "forge-std/Test.sol";
+import {SphinxVerifier as SphinxPlonkVerifier} from "sphinx-contracts/SphinxVerifier.sol";
+import "openzeppelin/access/Ownable.sol";
+
+struct SphinxProofFixture {
+    bytes proof;
+    bytes publicValues;
+    bytes32 vkey;
+}
+
+contract Wrapper is SphinxPlonkVerifier, Ownable(msg.sender) {
+    error ErrorUnexpectedLongestChainFixture();
+    error ErrorUnexpectedSpvFixture();
+    error ErrorUnexpectedSpvSubjectHash();
+    error ErrorConfirmationWorkIsSmallerThanThreshold();
+    error ErrorUnexpectedLayerHash();
+
+    // state
+    bytes32[] checkpoints;
+    uint256 confirmation_work_threshold;
+
+    constructor(uint256 confirmation_work_threshold_, bytes32[] memory checkpoints_) {
+        // Kadena's Chainweb protocol setting
+        require(checkpoints_.length == 5);
+
+        checkpoints = checkpoints_;
+        confirmation_work_threshold = confirmation_work_threshold_;
+    }
+
+    // FIXME: This function is only for testing purposes!!!
+    function set_head_checkpoint(bytes32 zero_checkpoint) public onlyOwner {
+        checkpoints[0] = zero_checkpoint;
+    }
+
+    // FIXME: This function is only for testing purposes!!!
+    function set_confirmation_work_threshold(uint256 threshold) public onlyOwner {
+        confirmation_work_threshold = threshold;
+    }
+
+    function rotate_checkpoints(bytes32 new_checkpoint) public onlyOwner {
+        // drop hash from [0] position, shift rest of hashes and write new checkpoint to the [4] position
+        checkpoints[0] = checkpoints[1];
+        checkpoints[1] = checkpoints[2];
+        checkpoints[2] = checkpoints[3];
+        checkpoints[3] = checkpoints[4];
+        checkpoints[4] = new_checkpoint;
```

Comment:
I feel this could be implemented with recursion to allow for any length of the `checkpoints` array.

Reply:
I think recursion is always a risk for OutOfGas in Solidity, so I've rewritten it using regular for loop